beacon-chain/state/state-native/v1/getters_block.go

```diff
@@ -50,50 +46,51 @@ func (b *BeaconState) latestBlockHeader() *ethpb.BeaconBlockHeader {
 
 // BlockRoots kept track of in the beacon state.
 func (b *BeaconState) BlockRoots() [][]byte {
-	if !b.hasInnerState() {
-		return nil
-	}
-	if b.state.BlockRoots == nil {
+	if b.blockRoots == nil {
 		return nil
 	}
 
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
-	return b.blockRoots()
+	rootsArr := b.blockRootsInternal()
+	roots := make([][]byte, len(rootsArr))
+	for i, r := range rootsArr {
+		tmp := r
+		roots[i] = tmp[:]
+	}
+	return roots
```

Comment:
This is quite a common pattern. Internal methods return custom types which are not assignable directly to slice arrays. If we did not external methods to conform to a single interface with the proto state, they would also return custom types and these copies would not be needed.

Reply:
Maybe as part of your tracking issue or another issue we can revisit this so that we can avoid allocating more memory for these roots. This alone is 262 Kb of data allocated, and I think we should do something about it. It can easily add up